### ISSUE TITLE
Specialize sum over arrays of static numbers

### ISSUE THREAD

**Issue by oschulz** [SciML/Static.jl#51]
We currently have

```julia
mapreduce(identity, +, fill(static(0), 5)) isa StaticInt{0}
```

but

```julia
sum(fill(static(0), 5)) isa Int
```

**Comment by Tokazama** [comment]
Yeah, that's not great. Is this something you've been running into or is it just a result of thorough testing?

**Comment by oschulz** [comment]
I plan to use static numbers for statistical weights in BAT.jl (I often have scenarios where samples may or may not be weighted, depending on sampling algorithm), so I played around a bit to make sure that operations like `sum(fill(static(...), ...))` and `sum(FillArrays.Fill(static(...), ...))` are efficient.

Somewhat related: JuliaArrays/FillArrays.jl#176

**Comment by chriselrod** [comment]
This is what `reduce_tup` is for.

**Comment by Tokazama** [comment]
If you have 10^5 unique `StaticInt` values then I'm not sure it's the right type. If what you're actually dealing with is a handful of unique statically known values, but they are assigned to 10^5 unique variables then you probably want a different data structure than `Vector`. 

That's not to say there's nothing we can do here to make the situation better, but it seems like one of those corner cases where an unintuitive or involved solution may be worse than just explicitly  telling people that they should avoid that pattern. Just my initial thoughts.

**Comment by oschulz** [comment]
> If you have 10^5 unique StaticInt values 

You can easily end up with that with genric code, e.g. starting from `FillArrays.Fill(static(1), 10^5)` (I would use to represent a static statistical weight for a large number of samples, for example). There are many operations that will turn a `FillArrays.Fill` into a standard `Array`.

**Comment by Tokazama** [comment]
Oh, so you're starting off with an ideal structure where the default is for everything to be `1` but then it gets converted when reducing? I can see how that would be easy to fall into.

I still think it could be problematic to implement a fix on our end. We'd have to somehow be aware of a dynamic number of iterations through a loop or awkwardly define `Base.sum(a::AbstractArray{StaticInt})`. But if someone has a clean solution I'm game.

**Comment by oschulz** [comment]
> Oh, so you're starting off with an ideal structure where the default is for everything to be 1 but then it gets converted when reducing? I can see how that would be easy to fall into.

Yes, exactly. I need lot's of genric code that handles samples, and they may all have weight 1 - in which case important optimizations can be made or they may have arbitrary integer of floating point weights. So it would seem natural to use something like `Ones{StaticInt}(n)` (n can be very large) or `Fill(static(1), n`. But even with certain possible specializations (see JuliaArrays/FillArrays.jl#177) one can still end up with an `AbstractArray{StaticInt}` after a while. So it would be important that standard operations like `sum`, `mapreduce` etc. function efficiently on arrays of static numbers, and ideally propagate the "staticness" of the element type.

**Comment by oschulz** [comment]
> I still think it could be problematic to implement a fix on our end.

Hm

```julia
A = fill(static(1), 10)
sum(A)
```

comes down to

```
mapreduce(identity, Base.add_sum, A)
```

I think the culprit may be

```
julia> x = static(1)
static(1)

julia> x + x
static(2)

julia> Base.add_sum(x, x)
2
```

This seems to be due to the return-type assertion in the definition of `Base.add_sum`

```julia
add_sum(x::Real, y::Real)::Real = x + y
```

Since

```julia
julia> foo(a::Real, b::Real) = a + b
foo (generic function with 1 method)

julia> foo(x, x)
static(2)

julia> bar(a::Real, b::Real)::Real = a + b
bar (generic function with 1 method)

julia> bar(x, x)
2
```

which I don't get, since `static(2) isa Real`.

**Comment by devmotion** [comment]
> The samples may all have weight 1 - in which case important optimizations can be made

A bit off-topic in this thread: Isn't this what StatsBase.UnitWeights is designed for?

**Comment by devmotion** [comment]
I didn't want to initiate a longer discussion here, I don't think it's the right place for it. Just briefly: I don't think it's possible, and I don't even think it's desirable, to get rid of `UnitWeights` - e.g., in many cases an `AbstractWeights` supertype is needed/desired and the weight type hierarchy and traits are needed for other weight types anyway. That doesn't rule out that for other applications a fix for the issue here would be desirable but for the weights example using `AbstractWeights` seems a bit more natural.

**Comment by Tokazama** [comment]
Like Chris pointed out, this also depends on knowing the size array in some cases to be type stable. It's hard to know what the best approach is for addressing this until we get some closure on https://github.com/JuliaLang/julia/pull/44538. 

This seems like the sort of thing that should be part of the methods internal interface with something like
```julia
function _add_sum(A::AbstractArray{StaticInt{N}}) where {N}
    if known_length(A) === nothing
        return length(A) * N
    else
        return StaticInt{known_length(A) * N}()
    end
end
```
because it's usually a bad idea to dispatch on element type of an abstract collection.

**Comment by oschulz** [comment]
Currently, `mapreduce` on static numbers seems to try to keep the staticness even if the array size isn't known, at huge performance cost:

```julia
julia> A = fill(static(4.2), 10^5);

julia> @btime mapreduce(sqrt, Base.add_sum, A)
  16.529 ms (0 allocations: 0 bytes)
```

Could we add a specialization like

```julia
function Base.mapreduce(f, ::typeof(Base.add_sum), A::AbstractVector{T}) where {T<:StaticFloat64}
    length(eachindex(A)) * T()
end
```

it would get us

```julia
julia> @btime mapreduce(sqrt, Base.add_sum, A)
  73.947 ns (1 allocation: 16 bytes)
420000.0
```

**Comment by oschulz** [comment]
> Could we add a specialization like ...

Darn, we'd get an ambiguity with StaticArrays with that:

```
julia> A = SVector(fill(static(4.2), 5)...)^C
julia> mapreduce(sqrt, Base.add_sum, A)
ERROR: MethodError: mapreduce(::typeof(sqrt), ::typeof(Base.add_sum), ::SVector{5, StaticFloat64{4.2}}) is ambiguous. Candidates:
  mapreduce(f, ::typeof(Base.add_sum), A::AbstractVector{T}) where T<:StaticFloat64
  mapreduce(f, op, a::StaticArray, b::StaticArray...; dims, init)
```
